python/ray/serialization.py

```diff
@@ -153,7 +153,15 @@ def id_deserializer(serialized_obj):
             return serialized_obj[0](*serialized_obj[1])
 
         def object_id_serializer(obj):
-            self.add_contained_object_id(obj)
+            if self.is_in_band_serialization():
+                self.add_contained_object_id(obj)
+            else:
+                # If this serialization is out-of-band (e.g., from a call to
+                # cloudpickle directly or captured in a remote function/actor),
+                # then pin the object for the lifetime of this worker by adding
+                # a local reference that won't ever be removed.
+                ray.worker.get_global_worker(
+                ).core_worker.add_object_id_reference(obj)
```

Comment:
Instead of modifying the serialization tracking, could we instead call get_and_clear_contained_object_ids() at the end of pickling a function definition, and add all of those to globals?

Reply:
I don't think that solution would be enough, since `ObjectID`s can also get arbitrarily pickled by other parts of the user's code.